### PR TITLE
2.7 Relocate Systemd Files

### PR DESCRIPTION
**Pull request by manadart**
### Checklist

 - [x] Checked if it requires a [pylibjuju](https://github.com/juju/python-libjuju) change?
 - [x] Added [integration tests](https://github.com/juju/juju/tree/develop/tests) for the PR?
 - [x] Added or updated [doc.go](https://discourse.jujucharms.com/t/readme-in-packages/451) related to packages changed?
 - [x] Do comments answer the question of why design decisions were made?

----

## Description of change

This patch changes the location of systemd unit files and start-up scripts.

Instead of writing each service definition and it's exec-start script into a service folder in `/lib/systemd/system` and using dbus (systemd) to link them into `/etc/systemd/system`, we write them directly to `/etc/systemd/system`. This means the exec-start script is co-located with its service definition and prefixed with the service name.

This change was required to support Focal Fossa (20.04) which has these two directories on different partitions, causing systemd to fail.

The linking step remains for now, though it effectively does nothing - the service definitions are in the default search path, so reloading the systemd daemon should be sufficient to have them loaded.

Subsequent patches will take care of any required upgrade paths, including upgrading the OS series.

## QA steps

- Bootstrap to Focal: `juju bootstrap lxd systemd-test-focal --bootstrap-series focal --debug --no-gui --force --config image-stream=daily`.
- SSH to the controller machine and observe the shutdown, database and machine services and scripts in `etc/systemd/system`.
- `juju add-machine --series focal`
- `juju deploy cs:ubuntu-13 -n 2 --to 0,0 --series focal --force`
- `juju ssh 0` and observe the service units/scripts are present in `/etc/systemd/system`.
- `juju remove-unit ubuntu/1` and observe that the service and script for the unit disappear.

## Documentation changes

None.

## Bug reference

https://bugs.launchpad.net/juju/+bug/1860432

### CODE REVIEW

**Comment by manadart** [comment]
[Fix for Focal and Systemd](https://trello.com/c/963WYqU9/2239-fix-for-focal-and-systemd)

File: service/agentconf.go [hunk 1]
```diff
@@ -227,13 +227,7 @@ func (s *systemdServiceManager) WriteSystemdAgents(
 
 		// Otherwise we need to manually ensure the service unit links.
 		svcFileName := svcName + ".service"
-		if err = os.Symlink(path.Join(systemd.LibSystemdDir, svcName, svcFileName),
```
Comment (manadart) [inline service/agentconf.go]: This method (WriteSystemdAgents) is all at sea. I will refactor it in a forthcoming patch.

**Comment by manadart** [comment]
$$merge$$

**Comment by manadart** [comment]
$$merge$$

**Comment by mitechie** [comment]
We've chatted with the lxd team and it doesn't seem to be on their end at
the moment. Our decision was to get enablement and then chat with the
kernel folks for guidance.

**Comment by howbazaar** [comment]
One thing that was mentioned here was to not bother with an external script at all, but instead have multiple ExecStart lines rather than an external script.

The "install" command is a better way to create the log file with the right permissions and ownership.

I think it would be worthwhile dropping the exec script if we can get all the behaviour into just the service file.

**Comment by manadart** [comment]
We can look at how we might achieve multi-command `ExecStart`, but as the commands stand, they use bash-isms that are not supported by the format.
See: https://www.freedesktop.org/software/systemd/man/systemd.service.html#Command%20lines

**Comment by manadart** [comment]
$$merge$$

**Pull request merged**